Get Vacations

PgrNascimento opened the pull request:
## Objective

Map Vacation List endpoint from Glass Factory API and resolve the #32 issue

## TODO

- [x] Add documentation
- [x] Add `get_vacations!` function

PgrNascimento committed:
Add Vacation struct

PgrNascimento committed:
Add get_vacations endpoint

This endpoint permits us to see a list of all vacations or all vacations
from an user. At this moment we inform the user by Query String, we
could pass by query string too, the start and end date and the vacation
type, but, the GlassFactory API don't filter if we use these args =(

PgrNascimento commented on lib/vacations.ex:
```diff
@@ -0,0 +1,23 @@
+    end
+  end
+
+  defp vacations_path(user_id) when (is_nil(user_id) or user_id == ""), do: "vacations"
```
This guard clause was the solution I've found to solve this problem. Guys, can you collaborate with me to solve this problem from another perspective and improve me knowledge with the language? #gratidao

samuelpordeus replied:
why `user_id` can come as `nil` or `""`? 🤔

PgrNascimento replied:
It's optional in the query string params. So, if you want all vacations from all users, the user_id can be an empty string and if you want all vacations from an specific user, you need inform the user_id.

I've answered you my friend?

britto replied:
Since it's an optional parameter like the others, we can instead accept a bag of params instead and keep it more flexible:

```elixir
def get_vacations(params \\ [], config \\ %{}) do
  ApiClient.get(vacations_path(user_id), params, config)
end

# That would require a tiny adaptation in the ApiClient itself, but I think it's worth it:
def get(resource, params \\ [], configuration) do
  config = Configuration.build(configuration)
  Tesla.get(tesla_client(config), resource, query: params)
end
```

PgrNascimento replied:
@britto Can you help me with the content of `params` variable? Is it a map or a string or something different? 😅

PgrNascimento replied:
I've search with @Cendrao and we've found that `params` variable is a list of map.

https://hexdocs.pm/tesla/Tesla.html#get/3

britto replied:
> Can you help me with the content of params variable? Is it a map or a string or something different?

Sure! I'd love to.

> we've found that params variable is a list of map

In fact, the query params type is a "keyword list-like" thing, where keys can be strings, in addition to atoms: https://hexdocs.pm/tesla/Tesla.Env.html#t:query/0

samuelpordeus commented on lib/vacations.ex:
```diff
@@ -0,0 +1,23 @@
+
+  @doc """
+
+  """
```
I think you forgot to remove this

PgrNascimento replied:
I will add documentation, how I would like feedback from you, I've pushed without all things the I would like to do in this PR. I've added a WIP flag :)

samuelpordeus commented on lib/vacations.ex:
```diff
@@ -0,0 +1,23 @@
+defmodule GlassFactoryApi.Vacations do
+  @moduledoc """
+
+  """
```
`@moduledoc false`

samuelpordeus commented on test/vacations/vacation_test.exs:
```diff
@@ -0,0 +1,33 @@
+defmodule GlassFactoryApi.Vacations.VacationTest do
+  use ExUnit.Case, async: true
```
try to remember to always use `@moduledoc false` if you are not writing any :P

samuelpordeus commented on test/vacations/vacation_test.exs:
```diff
@@ -0,0 +1,33 @@
+defmodule GlassFactoryApi.Vacations.VacationTest do
```
try to remember to write `@moduledoc false` if you are not writing documentation for a module =P

PgrNascimento replied:
TIL. Thanks <3<3<3

PgrNascimento committed:
Update get function

To permit the get function receive args for query string, I've update
the function to receive one more arg.

PgrNascimento committed:
Change user_id param to query_string param

We've updated the ApiClient.get to receive query string. This way, we
don't need pass a string with query string.

PgrNascimento committed:
Add documentation to get_vacations function

PgrNascimento committed:
Add cases when the request is not a success

PgrNascimento committed:
Add get_vacations! function

PgrNascimento committed:
Add documentation to get_vacations! function

Extra: update get_vacation documentation

PgrNascimento committed:
Add permissive list to query string params

Cendrao commented on lib/api_client.ex:
```diff
@@ -12,13 +12,14 @@ defmodule GlassFactoryApi.ApiClient do
-  @spec get(String.t(), map()) :: Tesla.Env.t()
-  def get(resource, configuration) do
+  @spec get(String.t(), [map()], map()) :: Tesla.Env.t()
+  def get(resource, query_string \\ [], configuration) do
```
My only question is about this method, you are changing its arity from `/2` to `/3` it shouldn't break other places like [members](https://github.com/Cendrao/glass_factory_api/blob/master/lib/members.ex#L33) or [projects](https://github.com/Cendrao/glass_factory_api/blob/master/lib/projects.ex#L30)? 

Even `query_string` has an optional value when calling it like:

```elixir
ApiClient.get("members, config)
```

Looking at the function definition `def get(resource, query_string \\ [], configuration)`, wouldn't the config be the 2nd parameter and configuration come `nil` and we get an error?

Cendrao replied:
OBS: The tests are passing and I did a manual test in the console which worked too, I'm very confused about it 😅

tegon replied:
It's because of the default value (`query_string \\ []`). Every time I see this pattern I find it very confusing too. Whenever possible, I'd leave the optional argument as the last one.

britto replied:
I'm 🆗 with the query params in the middle, at least it is closer to the `resource`, which is more related to it than the config. But no strong opinions on it, feel free to swap them, the impact is small bc this is an "internal" API. About the name, I'd prefer `params` over `query_string`, because at this point it is not a string yet, it will only become a string when Tesla processes it.

Cendrao replied:
I don't have strong feelings to change too, I think we can leave this way.

I just was trying to understand 😄 Thanks @tegon!

Cendrao reviewed (approved):
Great 🔝 

Left just one question which is bothering me 🙈

britto commented on lib/clients.ex:
```diff
@@ -28,7 +28,7 @@ defmodule GlassFactoryApi.Clients do
   @spec get_client(integer(), map) :: {atom(), Client.t() | String.t()}
   def get_client(client_id, config \\ %{}) do
-    with {:ok, %{status: 200, body: body}} <- ApiClient.get("clients/#{client_id}", config) do
+    with {:ok, %{status: 200, body: body}} <- ApiClient.get("clients/#{client_id}", [], config) do
```
We shouldn't need to make these changes, because `[]` is already the default value.

britto commented on lib/vacations.ex:
```diff
@@ -0,0 +1,143 @@
+defmodule GlassFactoryApi.Vacations do
+  @moduledoc """
+  Provides the methods to get vacations of all or a specific user.
```
```suggestion
  Provides functions to get vacations of all or a specific user.
```

britto commented on lib/vacations.ex:
```diff
@@ -0,0 +1,143 @@
+
+  @doc """
+  Return a tuple with `:ok` atom and an list of vacations.
+  or an tuple with `:error` and a string with the error description.
```
```suggestion
  or a tuple with `:error` and a string with the error description.
```

britto commented on lib/vacations.ex:
```diff
@@ -0,0 +1,143 @@
+
+  ## Examples
+
+  iex> GlassFactoryApi.Clients.get_vacations()
```
I think we need to indent this with 4 spaces in order to render properly in ex_doc.

britto replied:
Just ran `ex_doc` and confirmed: we do need to indent this.

britto commented on lib/vacations.ex:
```diff
@@ -0,0 +1,143 @@
+  Return a tuple with `:ok` atom and an list of vacations.
+  or an tuple with `:error` and a string with the error description.
+
+  It's possible apply some filters, like user_id, start and end to limit dates, and vacation_type
```
We can also use bullet lists to document each option individually. Take a look at [`String.split/3`](https://hexdocs.pm/elixir/String.html#split/3) for a good example.

britto commented on lib/vacations.ex:
```diff
@@ -0,0 +1,143 @@
+  end
+
+  defp filter_query_string(query_string) do
+    Keyword.take(query_string, [:start, :end, :user_id, :vacation_type])
```
Good usage of `take`! 👍

britto commented on lib/vacations.ex:
```diff
@@ -0,0 +1,143 @@
+
+  @spec get_vacations([], map()) :: {atom(), [Vacation.t() | String.t()]}
+  def get_vacations(query_string \\ [], config \\ %{}) do
+    with {:ok, %{status: 200, body: body}} <- ApiClient.get("vacations", filter_query_string(query_string), config) do
```
Same here, I'd favor `params` over `query_string`.

britto replied:
Actually, in `ApiClient` it makes sense to call it `params`, at this level here we should probably call it `filters` or `opts`.

britto commented on lib/vacations.ex:
```diff
@@ -0,0 +1,143 @@
+  @spec get_vacations([], map()) :: {atom(), [Vacation.t() | String.t()]}
+  def get_vacations(query_string \\ [], config \\ %{}) do
+    with {:ok, %{status: 200, body: body}} <- ApiClient.get("vacations", filter_query_string(query_string), config) do
+      {:ok, Enum.map(body, &Vacation.to_struct(&1))}
```
```suggestion
      {:ok, Enum.map(body, &Vacation.to_struct/1)}
```

britto commented on lib/vacations/vacation.ex:
```diff
@@ -0,0 +1,45 @@
+          daily_hours: String.t(),
+          start_date: String.t(),
+          end_date: String.t(),
+          days: String.t(),
```
String or integer?

PgrNascimento replied:
Yes, this should be an integer, but the Glass Factory API returns a string, and we decided to keep the same type. at least in this first version 😄

britto commented on lib/vacations/vacation.ex:
```diff
@@ -0,0 +1,45 @@
+  @type t() :: %Vacation{
+          id: String.t(),
+          upcoming: boolean(),
+          daily_hours: String.t(),
```
String or integer?

PgrNascimento replied:
Yes, this should be an integer, but the Glass Factory API returns a string, and we decided to keep the same type. at least in this first version 😄

britto replied:
Alright.

britto commented on test/vacations_test.exs:
```diff
@@ -0,0 +1,142 @@
+      assert vacations == Vacations.get_vacations!([], config)
+    end
+
+    test "raises an exception when something went wrong", %{bypass: bypass, config: config} do
```
```suggestion
    test "raises an exception when something goes wrong", %{bypass: bypass, config: config} do
```

paulo-silva reviewed (approved):
LGTM, don't forget to run `mix format`

:shipit: 🍬

PgrNascimento committed:
Apply PR suggestions

britto commented on lib/vacations.ex:
```diff
@@ -0,0 +1,168 @@
+
+  ## Arguments
+
+  - query_string - options to filter the vacations.
```
I think we need to indent this too. Let's check ex_doc's output.

PgrNascimento replied:
@britto

Here a print screen from `get_vacation` function, I like this way, shows to me that options are to `query_string` param, but I could see your opinion too. WDYT?

<img width="948" alt="Captura de Tela 2019-12-02 às 11 42 49" src="https://user-images.githubusercontent.com/13913773/69968381-0d410a80-14f9-11ea-8bb5-946ad56eea3a.png">

britto replied:
Yes, it works without indentation. However, indenting bullet lists with 2 spaces seems to be a standard adopted by the Elixir core team, so I'd follow them. Examples:

https://github.com/elixir-lang/elixir/blob/055526057ac9992cc1df87441357c4e56d45235d/lib/elixir/lib/string.ex#L327-L333
https://github.com/elixir-lang/elixir/blob/055526057ac9992cc1df87441357c4e56d45235d/lib/elixir/lib/string.ex#L75-L79

PgrNascimento committed:
Adjust format of the lines

PgrNascimento committed:
Adjust typo

PgrNascimento committed:
Merge branch 'master' of github.com:Cendrao/glass_factory_api into aon-list-vacations

* 'master' of github.com:Cendrao/glass_factory_api:
  Add mix format execution before mix test.
  Mix default parameter from list client function (#35)

britto commented on lib/api_client.ex:
```diff
@@ -12,13 +12,14 @@ defmodule GlassFactoryApi.ApiClient do
-  @spec get(String.t(), map()) :: Tesla.Env.t()
-  def get(resource, configuration) do
+  @spec get(String.t(), Keyword.t(), map()) :: Tesla.Env.t()
+  def get(resource, query_string \\ [], configuration) do
```
I'd call this `params` instead of `query_string`, because it is not a string yet at this point.

britto replied:
Or maybe just `query`

britto commented on lib/vacations.ex:
```diff
@@ -0,0 +1,169 @@
+  """
+
+  @spec get_vacations(Keyword.t(), map()) :: {atom(), [Vacation.t() | String.t()]}
+  def get_vacations(query_string \\ [], config \\ %{}) do
```
Same here, I'd rename `query_string` to either `params` or `query`.

britto replied:
Or `filters`

britto commented on lib/vacations.ex:
```diff
@@ -0,0 +1,169 @@
+  @doc """
+  Similar to get_vacation/2, returns a list of vacations or raises an error.
+
+  ## Arguments
```
```
  ## Params

  The supported filters are:

    * `user_id` (positive integer) - to return only the vacations from the user id specified.
    * `start` (date yyyy-mm-dd) - to filter the vacations from a period.
    * `end` (date yyyy-mm-dd) - to filter the vacations up until a period.

  If you specify a `start` date, you need to set an `end` date too.
```

PgrNascimento committed:
Apply @britto suggestions.

britto reviewed (approved):
Great job, @PgrNascimento!! 💖 :shipit:
